stan/math/opencl/matrix_cl.hpp

```diff
@@ -298,6 +299,28 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
     }
   }
 
+  template <typename Mat, std::enable_if_t<is_eigen_matrix<Mat>::value>...>
+  inline void write_buffer_(Mat&& A) {
+    cl::Event transfer_event;
+    cl::CommandQueue& queue = opencl_context.queue();
+    queue.enqueueWriteBuffer(this->buffer_cl_, opencl_context.in_order(), 0,
+                             sizeof(T) * A.size(), A.data(), nullptr,
+                             &transfer_event);
+    this->add_write_event(transfer_event);
+  }
+
+  template <typename Mat,
+            std::enable_if_t<!is_eigen_matrix<Mat>::value
+                             || std::is_rvalue_reference<Mat>::value>...>
+  inline void write_buffer_(Mat&& A) {
+    cl::Event transfer_event;
+    cl::CommandQueue& queue = opencl_context.queue();
+    queue.enqueueWriteBuffer(this->buffer_cl_, opencl_context.in_order(), 0,
```

Comment:
```suggestion
    queue.enqueueWriteBuffer(this->buffer_cl_, CL_TRUE, 0,
```
This change fixes the test I provided. But there are still other constructors and functions in `copy.hpp` that are affected by this bug.